src/ledger/result_writer.zig

```diff
@@ -110,58 +110,143 @@ pub const LedgerResultWriter = struct {
     /// agave: set_duplicate_confirmed_slots_and_hashes
     pub fn setDuplicateConfirmedSlotsAndHashes(
         self: *Self,
-        duplicate_confirmed_slot_hashes: []struct { Slot, Hash },
+        duplicate_confirmed_slot_hashes: []const struct { Slot, Hash },
     ) !void {
-        var write_batch = try self.db.initWriteBatch();
+        var setter = try self.setDuplicateConfirmedSlotsAndHashesIncremental();
+        defer setter.deinit();
+        errdefer setter.cancel();
+
         for (duplicate_confirmed_slot_hashes) |slot_hash| {
-            const slot, const frozen_hash = slot_hash;
-            const data = FrozenHashVersioned{ .current = FrozenHashStatus{
+            const slot, const hash = slot_hash;
+            try setter.addSlotAndHash(slot, hash);
+        }
+
+        try setter.commit();
+    }
+
+    /// Returns a struct which can be used to enact the same operation as `setDuplicateConfirmedSlotsAndHashes`, incrementally.
+    pub fn setDuplicateConfirmedSlotsAndHashesIncremental(
+        self: *Self,
+    ) !SetDuplicateConfirmedSlotsAndHashesIncremental {
+        return .{
+            .result_writer = self,
+            .write_batch = try self.db.initWriteBatch(),
+            .is_committed = false,
+        };
+    }
+    pub const SetDuplicateConfirmedSlotsAndHashesIncremental = struct {
+        result_writer: *Self,
+        write_batch: BlockstoreDB.WriteBatch,
+        is_committed: bool,
+
+        pub fn deinit(self: *SetDuplicateConfirmedSlotsAndHashesIncremental) void {
+            std.debug.assert(self.is_committed);
+            self.write_batch.deinit();
+        }
+
+        pub fn addSlotAndHash(
+            self: *SetDuplicateConfirmedSlotsAndHashesIncremental,
+            slot: Slot,
+            frozen_hash: Hash,
+        ) !void {
+            std.debug.assert(!self.is_committed);
+            const data: FrozenHashVersioned = .{ .current = .{
                 .frozen_hash = frozen_hash,
                 .is_duplicate_confirmed = true,
             } };
-            try write_batch.put(schema.bank_hash, slot, data);
+            try self.write_batch.put(schema.bank_hash, slot, data);
+        }
+
+        pub fn cancel(self: *SetDuplicateConfirmedSlotsAndHashesIncremental) void {
+            std.debug.assert(!self.is_committed);
+            self.is_committed = true;
```

Comment:
Should `is_committed` be set to true on cancel?  especially given it is called in an  `errdefer`.

Reply:
It has to be, otherwise the assertion in the deinit will be tripped on the error path. That's why this function exists to be errdefer'd.

Reply:
I went ahead and elaborated further documentation for these items, does this resolve your doubts? 67b3ec99e92caf85b87ce1972b0f4133f6fa3ffc

Reply:
Yup. The rename + the comments is definitely an improvement.